File: apps/api/src/serializers/entries.ts
import type {
  FeedbackNoteDto,
  IssueEntryDto,
  NoteDto,
  TaskEntryDto,
} from '@onboarding-diary/shared';

import type { FeedbackNote, IssueEntry, Note, TaskEntry } from '../generated/prisma/client.js';

function calendarDate(value: Date): string {
  return value.toISOString().slice(0, 10);
}

export function toTaskDto(task: TaskEntry): TaskEntryDto {
  return {
    id: task.id,
    ownerId: task.ownerId,
    entryDate: calendarDate(task.entryDate),
    title: task.title,
    description: task.description,
    category: task.category,
    status: task.status,
    priority: task.priority,
    createdAt: task.createdAt.toISOString(),
    updatedAt: task.updatedAt.toISOString(),
  };
}

export function toIssueDto(issue: IssueEntry): IssueEntryDto {
  return {
    id: issue.id,
    ownerId: issue.ownerId,
    entryDate: calendarDate(issue.entryDate),
    title: issue.title,
    description: issue.description,
    severity: issue.severity,
    status: issue.status,
    resolutionNotes: issue.resolutionNotes,
    createdAt: issue.createdAt.toISOString(),
    updatedAt: issue.updatedAt.toISOString(),
  };
}

export function toFeedbackDto(feedback: FeedbackNote): FeedbackNoteDto {
  return {
    id: feedback.id,
    ownerId: feedback.ownerId,
    entryDate: calendarDate(feedback.entryDate),
    subject: feedback.subject,
    type: feedback.type,
    details: feedback.details,
    createdAt: feedback.createdAt.toISOString(),
    updatedAt: feedback.updatedAt.toISOString(),
  };
}

export type NoteWithTags = Note & { tags: { tag: { name: string } }[] };

export function toNoteDto(note: NoteWithTags): NoteDto {
  return {
    id: note.id,
    ownerId: note.ownerId,
    entryDate: calendarDate(note.entryDate),
    title: note.title,
    content: note.content,
    tags: note.tags.map((join) => join.tag.name).sort(),
    createdAt: note.createdAt.toISOString(),
    updatedAt: note.updatedAt.toISOString(),
  };
}

export { calendarDate as toCalendarDateString };

File: backend/src/OnboardingDiary.Api/Domain/Entries.cs
using System.Text.Json.Serialization;

namespace OnboardingDiary.Api.Domain;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum TaskCategory
{
    Training,
    Setup,
    Meeting,
    ProjectWork,
    Documentation,
    Other
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum TaskStatus
{
    NotStarted,
    InProgress,
    Completed,
    Blocked
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum TaskPriority
{
    Low,
    Medium,
    High
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum IssueSeverity
{
    Low,
    Medium,
    High,
    Critical
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum IssueStatus
{
    Open,
    InProgress,
    Resolved,
    Closed
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum FeedbackType
{
    Positive,
    Suggestion,
    Concern
}

/// Fields shared by every diary entry type.
public abstract class DiaryEntry
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid UserId { get; set; }
    public User? User { get; set; }
    public DateOnly Date { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;
}

public class TaskEntry : DiaryEntry
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public TaskCategory Category { get; set; }
    public TaskStatus Status { get; set; }
    public TaskPriority Priority { get; set; }
}

public class IssueEntry : DiaryEntry
{
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public IssueSeverity Severity { get; set; }
    public IssueStatus Status { get; set; }
    public string? ResolutionNotes { get; set; }
}

public class FeedbackNote : DiaryEntry
{
    public string Subject { get; set; } = string.Empty;
    public FeedbackType Type { get; set; }
    public string Details { get; set; } = string.Empty;
}

public class Note : DiaryEntry
{
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public List<string> Tags { get; set; } = [];
}
